kafka/dispatcher.go

```diff
@@ -0,0 +1,133 @@
+//go:generate mockgen -destination=./$GOFILE.mock_test.go -package=$GOPACKAGE github.com/Shopify/sarama SyncProducer
+package kafka
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/italolelis/outboxer"
+)
+
+var errInvalidProducerConfig = errors.New("invalid kafka producer config")
+
+const (
+	Topic     = "topic"
+	MetaData  = "metadata"
+	Partition = "partition"
+)
+
+type saramaDispatcher struct {
+	sarama.SyncProducer
+}
+
+func newSaramaDispatcher(client sarama.Client) (*saramaDispatcher, error) {
+	err := checkProducerConfig(client.Config())
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := sarama.NewSyncProducerFromClient(client)
+	if err != nil {
+		return nil, err
+	}
+
+	return &saramaDispatcher{p}, nil
+}
+
+func (d *saramaDispatcher) Dispatch(ctx context.Context, msg *outboxer.OutboxMessage) error {
+	errChan := make(chan error)
+
+	go func() {
+		errChan <- d.dispatchMessage(msg)
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (d *saramaDispatcher) dispatchMessage(message *outboxer.OutboxMessage) error {
+	producerMsg, err := buildSaramaProducerMessage(*message)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = d.SendMessage(&producerMsg)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func buildSaramaProducerMessage(message outboxer.OutboxMessage) (sarama.ProducerMessage, error) {
+	producerMsg := sarama.ProducerMessage{
+		Headers: make([]sarama.RecordHeader, len(message.Headers)),
+	}
+
+	opts := message.Options
+	topic, ok := opts[Topic]
+
+	if !ok {
+		return producerMsg, fmt.Errorf("%w: %s", errKafkaOptionMandatory, Topic)
+	}
+
+	producerMsg.Topic, ok = topic.(string)
+
+	if !ok || producerMsg.Topic == "" {
+		return producerMsg, fmt.Errorf(optionTypeErrFmt, errKafkaOptionType, Topic, "str")
+	}
+
+	if data, ok := opts[MetaData]; ok {
+		producerMsg.Metadata = data
+	}
+
+	if data, ok := opts[Partition]; ok {
+		if producerMsg.Partition, ok = data.(int32); !ok {
+			return producerMsg, fmt.Errorf(optionTypeErrFmt, errKafkaOptionType, Partition, int32(1))
+		}
+	} else {
+		producerMsg.Key = sarama.StringEncoder(strconv.FormatInt(message.ID, 10))
+	}
+
+	i := 0
+
+	for key, val := range message.Headers {
+		v, ok := val.([]byte)
+		if !ok {
+			return producerMsg, fmt.Errorf(optionTypeErrFmt, errKafkaOptionType, "Headers", map[string][]byte{})
+		}
+
+		producerMsg.Headers[i] = sarama.RecordHeader{
+			Key:   []byte(key),
+			Value: v,
+		}
```

Comment:
I already preallocated memory required for `Headers` to avoid new allocations during append.

```
producerMsg := sarama.ProducerMessage{
		Headers: make([]sarama.RecordHeader, len(message.Headers)),
	}
```